src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb

```diff
@@ -491,47 +493,79 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Select
         End Function
 
-        Public Overrides Function GetHashCode() As Integer
-            Dim _hash As Integer = OriginalDefinition.GetHashCode()
+        Public NotOverridable Overrides Function GetHashCode() As Integer
+            Dim hash As Integer = OriginalDefinition.GetHashCode()
             If Me._substitution.WasConstructedForModifiers() Then
-                Return _hash
+                Return hash
             End If
 
-            _hash = Hash.Combine(ContainingType, _hash)
+            hash = Roslyn.Utilities.Hash.Combine(ContainingType, hash)
 
             ' There is a circularity problem here with alpha-renamed type parameters.
             ' Calculating GetHashCode for them calls back into container's GetHashCode.
-            ' Do not ask for hash code of type arguments here, derived classes 
-            ' override this function and do that when appropriate. 
-            Return _hash
-        End Function
+            If Me IsNot Me.ConstructedFrom Then
+                For Each typeArgument In TypeArgumentsNoUseSiteDiagnostics
+                    hash = Roslyn.Utilities.Hash.Combine(typeArgument, hash)
+                Next
+            End If
 
-        Public MustOverride Overrides Function Equals(obj As Object) As Boolean
+            Return hash
+        End Function
 
-        ''' <summary>
-        ''' Compare SubstitutedNamedTypes with no regard to type arguments.
-        ''' </summary>
-        Private Function EqualsWithNoRegardToTypeArguments(Of T As SubstitutedNamedType)(other As T) As Boolean
+        Public NotOverridable Overrides Function Equals(other As TypeSymbol, comparison As TypeCompareKind) As Boolean
+            If other Is Me Then
+                Return True
+            End If
 
             If other Is Nothing Then
                 Return False
             End If
 
+            If (comparison And TypeCompareKind.AllIgnoreOptionsForVB) = 0 AndAlso
+               Not Me.GetType().Equals(other.GetType()) Then
+                Return False
+            End If
+
+            Dim otherTuple = TryCast(other, TupleTypeSymbol)
+            If otherTuple IsNot Nothing Then
```

Comment:
Where do the type arguments for tuple types get validated? The validation for a normal substituted type arguments are below this check hence they're skipped here. The implementation of `TupleTypeSymbol` equality doesn't appear to check them.

Reply:
> Where do the type arguments for tuple types get validated? The validation for a normal substituted type arguments are below this check hence they're skipped here. The implementation of TupleTypeSymbol equality doesn't appear to check them. 

The implementation of TupleTypeSymbol equality compares underlying types, that takes care of comparing type arguments for them. In VB TupleTypeSymbol itself doesn't have any type arguments. 

---
In reply to: [533616189](https://github.com/dotnet/roslyn/pull/49569#discussion_r533616189) [](ancestors = 533616189)